### DISCUSSION TITLE
Cannot implement Sub-Dependencies by combine 2 dependencies

### DISCUSSION THREAD

**Discussion by ammarsaf** [fastapi/fastapi#11314] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def car_params(name: Union[str, None], brand: Union[str, None]):
    return {"name": name, "brand": brand}
def motor_params(name_m: Union[str, None], brand_m: Union[str, None]):
    return {"name_m": name_m, "brand_m": brand_m}

# nested schema
def transport_param(
    car: Annotated[dict, Depends(car_params)],
    motor: Annotated[dict, Depends(motor_params)],
):
    if not car:
        return motor
    else:
        return car

@app.post("/cardetail/3")
def get_car(
    info: Annotated[dict, Depends(transport_param)],
):
    return info
```

### Description

0I want to combine 2 Dependency in FastAPI to implement [Nested Model](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/)

I try running the endpoint on Docs, but it shows the loading too long! Feels something wrong here.

Are the implementation is wrong?

Thanks.

### Operating System

Linux

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

3.9.13

**Comment by YuriiMotov** [comment, marked as answer]
**Variant #1**

You need to specify the default values to make these parameters not required:
```
def car_params(name: Union[str, None] = None, brand: Union[str, None] = None):
    return {"name": name, "brand": brand}
def motor_params(name_m: Union[str, None] = None, brand_m: Union[str, None] = None):
    return {"name_m": name_m, "brand_m": brand_m}
```
Also, there is a bug in FastAPI, and your optional query parameters will appear without type annotations in Swagger. To fox it, use `SkipJsonSchema`:
```
def car_params(name: Union[str, SkipJsonSchema[None]] = None, brand: Union[str, SkipJsonSchema[None]] = None):
    return {"name": name, "brand": brand}
def motor_params(name_m: Union[str, SkipJsonSchema[None]] = None, brand_m: Union[str, SkipJsonSchema[None]] = None):
    return {"name_m": name_m, "brand_m": brand_m}
```
And then you should change the condition in the `transport_param` to check that the `car` dict is empty or contains `None`'s (thanks to @omarcruzpantoja):
```
    if not any(car.values()):
        return motor
    else:
        return car
```

**So, final solution is:**
```
from typing import Annotated, Union
from fastapi import Depends, FastAPI
from pydantic.json_schema import SkipJsonSchema

def car_params(name: Union[str, SkipJsonSchema[None]] = None, brand: Union[str, SkipJsonSchema[None]] = None):
    return {"name": name, "brand": brand}
def motor_params(name_m: Union[str, SkipJsonSchema[None]] = None, brand_m: Union[str, SkipJsonSchema[None]] = None):
    return {"name_m": name_m, "brand_m": brand_m}

# nested schema
def transport_param(
    car: Annotated[dict, Depends(car_params)],
    motor: Annotated[dict, Depends(motor_params)],
):
    if not any(car.values()):
        return motor
    else:
        return car

app = FastAPI()

@app.post("/cardetail/3")
def get_car(
    info: Annotated[dict, Depends(transport_param)],
):
    return info
```

**Variant 2**

If you want to make one of these parameters set required, you can implement it using Pydantic's [discriminated unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions).
I managed to create example only for Body parameters, but I think it's also possible to use Query parameters:
```
from typing import Annotated, Union

from fastapi import Body, FastAPI
from pydantic import BaseModel, Discriminator, Tag

app = FastAPI()

class CarParams(BaseModel):
    name: str
    brand: str

class MotorParams(BaseModel):
    name_m: str
    brand_m: str

def get_discriminator_value(v: dict) -> str:
    return "car" if v.get('name', None) else "motor"

@app.post("/cardetail/3")
def get_car(
    info: Annotated[
        Union[
            Annotated[CarParams, Tag("car")],
            Annotated[MotorParams, Tag("motor")]
        ],
        Body(discriminator=Discriminator(get_discriminator_value))
    ],
):
    return info
```

**Reply by ammarsaf** [reply]
Thank you for the reply. I tried your first solution, it returned a response successfully when I request on `car` param query, however, when I pass argument to `motor` param query, it returned 

```
{
  "name": null,
  "brand": null
}
```
which is belongs to the `car` response body, not the `motor` response body. 

The 2nd solution (using schema) is good, however I intent to use `Depends` to implement this composite param.

**Reply by omarcruzpantoja** [reply]
To further expand on @YuriiMotov solution  and solve the issue you're having with `nulls`. The reason is because the `car` variable will contain data. 
The data will look like: 

``` 
{
  "name": null,
  "brand": null
}
```

Since `car` WILL be a dictionary with data it will be `Truthy`, and return `car` instead of returning `motor`

---

Tldr; 

You'd need to update 
```
    if not car:
        return motor
    else:
        return car
```

to something like 

```
    if not any(car.values()): # This will verify if any of the values  in `car` have `Truthy` data. ( modify it to your needs to ensure the behavior you'd want )
        return motor
    else:
        return car
``` 

Edit: 
fixing some typos

**Reply by omarcruzpantoja** [reply]
Final solution could take the form of: 

```
def car_params(name: str | None = None, brand: str | None = None):
    return {"name": name, "brand": brand}


def motor_params(name_m: str | None = None, brand_m: str | None = None):
    return {"name_m": name_m, "brand_m": brand_m}


# nested schema
def transport_param(
    car: Annotated[dict, Depends(car_params)],
    motor: Annotated[dict, Depends(motor_params)],
):
    if not any(car.values()):
        return motor
    else:
        return car
```

Note: Used `str | None` instead of `Union[str, None]`

**Reply by YuriiMotov** [reply]
Great! Thanks!
I would only add using `SkipJsonSchema` to fix OpenAPI schema.
I've updated my answer

**Reply by ammarsaf** [reply]
Nice debugs! @omarcruzpantoja  . I just overlooked on the Truthy part in Python. 

Thank you to @YuriiMotov for the solution. Will mark yours as answer!
